File: example/UILabel+FormattedText/ViewController.m
//
//  ViewController.m
//  UILabel+FormattedText
//
//  Created by Joao Costa on 3/1/13.
//  Copyright (c) 2013 none. All rights reserved.
//

#import "ViewController.h"
#import "UILabel+FormattedText.h"

#define SEP 20
#define HEI 40

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    UIFont *bold = [UIFont fontWithName:@"Courier-Bold" size:14];
    UIFont *italic = [UIFont fontWithName:@"Courier-Oblique" size:14];
    
    // Colors
    UILabel *label0 = [self labelNo: 0];
    [label0 setText:@"This is a text with\nRED and CYAN"];
    [label0 setTextColor:[UIColor redColor] range:NSMakeRange(20, 3)];
    [label0 setTextColor:[UIColor cyanColor] range:NSMakeRange(28, 4)];

    // Fonts
    UILabel *label1 = [self labelNo: 1];
    [label1 setText:@"This is a text with\nBOLD and ITALIC"];
    [label1 setFont:bold range:NSMakeRange(20, 4)];
    [label1 setFont:italic range:NSMakeRange(29, 6)];
    
    // Colors AND Fonts
    UILabel *label2 = [self labelNo: 2];
    [label2 setText:@"This is a text with\nBOLD-RED and ITALIC-CYAN"];
    [label2 setTextColor:[UIColor redColor] range:NSMakeRange(20, 8)];
    [label2 setFont:bold range:NSMakeRange(20, 8)];
    [label2 setTextColor:[UIColor cyanColor] range:NSMakeRange(33, 11)];
    [label2 setFont:italic range:NSMakeRange(33, 11)];
    
    // Colors AND Fonts AND Underline
    UILabel *label3 = [self labelNo: 3];
    [label3 setText:@"This is a text with\nRED-UNDERLINE and CYAN-UNDERLINE"];
    [label3 setTextColor:[UIColor redColor] range:NSMakeRange(20, 13)];
    [label3 setTextUnderLine:[UIColor cyanColor] range:NSMakeRange(20, 13)];
    [label3 setTextColor:[UIColor cyanColor] range:NSMakeRange(38, 14)];
    [label3 setTextUnderLine:[UIColor redColor] range:NSMakeRange(38, 14)];
    
    // Colors and Remove Underline
    UILabel *label4 = [self labelNo: 4];
    [label4 setText:@"This is an underlined text with\nRED and CYAN"];
    [label4 setTextUnderLine:[UIColor yellowColor] range:NSMakeRange(0, 44)];
    [label4 setTextColor:[UIColor redColor] range:NSMakeRange(32, 3)];
    [label4 setTextWithoutUnderLineInRange:NSMakeRange(32, 3)];
    [label4 setTextColor:[UIColor cyanColor] range:NSMakeRange(40, 4)];
    [label4 setTextWithoutUnderLineInRange:NSMakeRange(40, 4)];
    
    // Changes after occurence
    UILabel *label5 = [self labelNo:5];
    [label5 setText:@"This is a text with\nRED-BOLD after 'RED-BOLD'"];
    [label5 setTextColor:[UIColor redColor] afterOccurenceOfString:@"RED-BOLD"];
    [label5 setFont:bold afterOccurenceOfString:@"RED-BOLD"];

    // Changes before occurence
    UILabel *label6 = [self labelNo:6];
    [label6 setText:@"This is a text with\nRED-BOLD before 'RED-BOLD'"];
    [label6 setTextColor:[UIColor redColor] beforeOccurenceOfString:@"RED-BOLD"];
    [label6 setFont:bold beforeOccurenceOfString:@"RED-BOLD"];
    
    // Changes after and before occurence
    UILabel *label7 = [self labelNo:7];
    [label7 setText:@"This is a text with\nRED from 'RED' to period mark."];
    [label7 setTextColor:[UIColor redColor] fromOccurenceOfString:@"RED" toOccurenceOfString:@"."];
    
    // Changes of precise text
    UILabel *label8 = [self labelNo:8];
    [label8 setText:@"This is a text with\nRED and RED-UNDERLINE using search."];
    [label8 setTextColor:[UIColor redColor] String:@"RED"];
    [label8 setTextColor:[UIColor redColor] String:@"RED-UNDERLINE" underLineColor:[UIColor cyanColor]];
}

- (UILabel *)labelNo:(CGFloat)index {
    UILabel *label = [[UILabel alloc] init];
    [label setFrame:CGRectMake(0, SEP + (HEI + SEP) * index, self.view.frame.size.width, HEI)];
    [label setBackgroundColor:[UIColor clearColor]];
    [label setTextAlignment:NSTextAlignmentCenter];
    [label setTextColor:[UIColor yellowColor]];
    [label setFont:[UIFont fontWithName:@"Courier" size:14]];
    [label setNumberOfLines:0];
    [self.view addSubview:label];
    
    return label;
}

- (BOOL)prefersStatusBarHidden {
    return YES;
}

@end

File: example/UILabel+FormattedText/ViewController.swift
//
//  ViewController.swift
//  UILabel+FormattedText
//
//  Created by João Costa on 16/09/15.
//
//

import UIKit

class ViewController: UIViewController {
    private let sep: CGFloat = 20
    private let hei: CGFloat = 40

    private let red = UIColor.redColor()
    private let cyan = UIColor.cyanColor()
    
    private let bold = UIFont(name: "Courier-Bold", size: 15)!
    private let italic = UIFont(name: "Courier-Oblique", size: 15)!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        view.backgroundColor = UIColor.blackColor()
        
        let title = titleAt(0)
        title.font = UIFont(name: "Courier-Bold", size: 25)
        title.textColor = UIColor.yellowColor()
        title.text = "UILabel+FormattedText"
        title.setTextUnderline(UIColor.yellowColor(), range: title.fullRange())

        // RANGES
        
        let title1 = titleAt(1)
        title1.text = "Formatting w/ Ranges"
        
        // Colors
        let label11 = labelAt(2)
        label11.text = "This is a text with\nRED and CYAN"
        label11.setTextColor(red, range: NSMakeRange(20, 3))
        label11.setTextColor(cyan, range: NSMakeRange(28, 4))
        
        // Fonts
        let label12 = labelAt(3)
        label12.text = "This is a text with\nBOLD and ITALIC"
        label12.setFont(bold, range: NSMakeRange(20, 4))
        label12.setFont(italic, range: NSMakeRange(29, 6))

        // Underline
        let label13 = labelAt(4)
        label13.text = "This is a text with\nFULL UNDERLINE and UNDERLINE BY WORD"
        label13.setTextUnderline(red, range: NSMakeRange(20, 14))
        label13.setTextUnderline(cyan, range: NSMakeRange(39, 17), byWord: true)
        
        // Remove Underline
        let label14 = labelAt(5)
        label14.text = "This is an underlined text with a NON UNDERLINED part"
        label14.setTextUnderline(cyan, range: label14.fullRange())
        label14.setTextWithoutUnderline(range: NSMakeRange(33, 16))
        
        // STRING MATCHERS
        
        let title2 = titleAt(6)
        title2.text = "Formatting w/ String Matchers"
        
        // After & Before
        let label21 = labelAt(7)
        label21.text = "Bold after THIS. Red before THAT. Underline after *THIS* and before *THAT*"
        label21.setFont(bold, range: label21.rangeAfter("THIS"))
        label21.setTextColor(red, range: label21.rangeBefore("THAT"))
        label21.setTextUnderline(cyan, range: label21.rangeAfter("*THIS*", before: "*THAT*"))
        
        // From & To
        let label22 = labelAt(8)
        label22.text = "Bold from THIS. Red to THAT. Underline from *THIS* and to *THAT*"
        label22.setFont(bold, range: label22.rangeFrom("THIS"))
        label22.setTextColor(red, range: label22.rangeTo("THAT"))
        label22.setTextUnderline(cyan, range: label22.rangeFrom("*THIS*", to: "*THAT*"))
        
        // String
        let label23 = labelAt(9)
        label23.text = "This is a text with\nRED, BOLD and UNDERLINE"
        label23.setTextColor(red, range: label23.rangeOf("RED"))
        label23.setFont(bold, range: label23.rangeOf("BOLD"))
        label23.setTextUnderline(cyan, range: label23.rangeOf("UNDERLINE"))
    }
    
    override func prefersStatusBarHidden() -> Bool {
        return true
    }
    
    private func titleAt(index: Int) -> UILabel {
        let y = sep + (hei + sep) * CGFloat(index)
        
        let title = UILabel(frame: CGRect(x: 0, y: y, width: view.frame.width, height: hei))
        title.textColor = UIColor.whiteColor()
        title.font = UIFont(name: "Courier-Bold", size: 18)
        title.textAlignment = .Center
        view.addSubview(title)
        
        return title
    }
    
    private func labelAt(index: Int) -> UILabel {
        let y = sep + (hei + sep) * CGFloat(index)
        
        let label = UILabel(frame: CGRect(x: 0, y: y, width: view.frame.width, height: hei))
        label.textAlignment = .Center
        label.textColor = UIColor.yellowColor()
        label.font = UIFont(name: "Courier", size: 15)
        label.numberOfLines = 0
        view.addSubview(label)
        
        return label
    }
}
